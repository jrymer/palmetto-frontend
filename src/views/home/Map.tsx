import React from 'react';
import leaflet from 'leaflet';
import styled from 'styled-components';

import { Coordinates } from '../../types';

const Container = styled.div`
  height: -webkit-fill-available;
  min-width: 40%;
  width: var(--col-width);
  grid-area: Map;
  border: solid thin ${(props) => props.theme.colors.offsetGrey};
  margin: ${(props) => props.theme.margin};
`;

interface MapProps {
  coords: Coordinates;
}

const tileLayerConf: leaflet.TileLayerOptions = {
  attribution:
    'Map data &copy; <a href="https://www.openstreetmap.org/copyright">OpenStreetMap</a> contributors, Imagery Â© <a href="https://www.mapbox.com/">Mapbox</a>',
  maxZoom: 14,
  id: 'mapbox/streets-v11',
  tileSize: 512,
  zoomOffset: -1,
  accessToken: process.env.REACT_APP_MAP_BOX_API_KEY,
};

/**
 * Returns a leaflet map of the weather location searched
 *
 * @param {Coordinates} { coords } LAT LON set to be plotted
 * @return A leaflet map
 */
const Map: React.FC<MapProps> = ({ coords }) => {
  const ref = React.useRef(true);

  /**
   * Initialize the map
   */
  React.useEffect(() => {
    if (ref.current && coords) {
      const { lat, lon } = coords;
      const map = leaflet.map('map').setView([lat, lon], 13);
      leaflet
        .tileLayer(
          `https://api.mapbox.com/styles/v1/${tileLayerConf.id}/tiles/{z}/{x}/{y}?access_token=${tileLayerConf.accessToken}`,
          {
            ...tileLayerConf,
          }
        )
        .addTo(map);
      const precip = leaflet.tileLayer(
        'https://tile.openweathermap.org/map/precipitation_new/{z}/{x}/{y}.png?appid=471ff37832bebfa6f9a5a94a2c0c4b34',
        { maxZoom: 14 }
      );
      const temp = leaflet.tileLayer(
        'https://tile.openweathermap.org/map/temp_new/{z}/{x}/{y}.png?appid=471ff37832bebfa6f9a5a94a2c0c4b34',
        { maxZoom: 14 }
      );
      const clouds = leaflet.tileLayer(
        'https://tile.openweathermap.org/map/clouds_new/{z}/{x}/{y}.png?appid=471ff37832bebfa6f9a5a94a2c0c4b34',
        { maxZoom: 14 }
      );
      const wind = leaflet.tileLayer(
        'https://tile.openweathermap.org/map/wind_new/{z}/{x}/{y}.png?appid=471ff37832bebfa6f9a5a94a2c0c4b34',
        { maxZoom: 14 }
      );

      const weatherLayers = {
        Precipitation: precip,
        Temperature: temp,
        'Wind Speed': wind,
        'Cloud coverage': clouds,
      };

      leaflet.control.layers(weatherLayers).addTo(map);

      leaflet
        .marker([lat, lon], {
          title: 'Location weather data was pulled from',
        })
        .addTo(map);
    }
    return () => {
      ref.current = false;
    };
  }, [coords]);

  return (
    <Container>
      <div id="map" />
    </Container>
  );
};

export default Map;
