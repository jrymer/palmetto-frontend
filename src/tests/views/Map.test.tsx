import React from 'react';

import Map from '../../views/home/Map';
import { createSnapshot } from '../testWrapper';

describe('Map tests', () => {
  it('Map snapshot is valid with base props', () => {
    const snapshot = createSnapshot(<Map coords={{ lat: 1, lon: 1 }} />);
    expect(snapshot).toMatchSnapshot();
  });
});
