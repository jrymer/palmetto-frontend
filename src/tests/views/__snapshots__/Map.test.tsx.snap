// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Map tests Map snapshot is valid with base props 1`] = `
<div
  className="sc-bczRLJ igATvh"
>
  <div
    id="map"
  />
</div>
`;
